torchrl/data/tensordict/tensordict.py

```diff
@@ -1991,7 +1995,8 @@ def set(
         and if the key already exists, set will call set_ (in place setting).
         """
         if self.is_locked:
-            raise RuntimeError("Cannot modify immutable TensorDict")
+            if not inplace or key not in set(self.keys()):
```

Comment:
you don't need to create a set every time :)
The idea is this, say you have a function that does this

```
if key in self.keys():
   do_stuff()
elif key is not in self.keys() and foo:
    val = bar
```
then you're iterating twice over the `keys()` iterator, which can be costly. Then we create a set once and avoid calling the iterator.
If `keys()` is called just once it's actually faster to leave it as it is (because creating a set is not free and the condition `if key in self.keys()` will stop whenever a matching keys is found, which can happen before all keys have been scanned)